File: 백준/Gold/1197. 최소 스패닝 트리/최소 스패닝 트리.py
import sys
input = sys.stdin.readline
sys.setrecursionlimit(10**6)

v,e = map(int,input().split())
edges = []

for _ in range(e):
    a,b,c = map(int,input().split())
    edges.append((a,b,c))

edges.sort(key = lambda x:x[2])

parent = [i for i in range(v+1)]

def getParent(x):
    if parent[x] == x:
        return x
    else:
        parent[x] = getParent(parent[x])
        return parent[x]

def sameParent(a,b):
    return getParent(a) == getParent(b)

def unionParent(a,b):
    a = getParent(a)
    b = getParent(b)

    if a < b:
        parent[b] = a
    else:
        parent[a] = b

answer = 0

for a,b,cost in edges:
    if not sameParent(a,b):
        unionParent(a,b)
        answer += cost

print(answer)

File: SWEA/D4/3124. 최소 스패닝 트리/최소 스패닝 트리.java
import java.io.*;
import java.util.*;

// # 크루스칼
public class Solution {

    public static class Edge implements Comparable<Edge>{
        int a;
        int b;
        int cost;

        Edge(int a, int b, int cost) {
            this.a = a;
            this.b = b;
            this.cost = cost;
        }

        @Override
        public int compareTo(Edge o) {
            return this.cost - o.cost;
        }
    }

    static int[] parent;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        int TC = Integer.parseInt(br.readLine());

        for (int testCase = 1; testCase <= TC; testCase++) {
            StringTokenizer st = new StringTokenizer(br.readLine(), " ");
            int v = Integer.parseInt(st.nextToken());
            int e = Integer.parseInt(st.nextToken());

            Edge[] edges = new Edge[e];
            for (int i = 0; i < e; i++) {
                st = new StringTokenizer(br.readLine(), " ");
                int a = Integer.parseInt(st.nextToken());
                int b = Integer.parseInt(st.nextToken());
                int c = Integer.parseInt(st.nextToken());

                edges[i] = new Edge(a, b, c);
            }

            Arrays.sort(edges);

            parent = new int[v+1];
            for (int i = 0; i <= v; i++) {
                parent[i] = i;
            }

            int count = 0;
            long answer = 0;
            for (Edge edge : edges) {
                int rootA = find(edge.a);
                int rootB = find(edge.b);

                if (rootA != rootB) {
                    count++;
                    answer += edge.cost;
                    union(rootA, rootB);

                    if (count == v-1) break;
                }
            }

            sb.append("#").append(testCase).append(" ").append(answer).append("\n");
        }
        System.out.print(sb);
    }

    public static int find(int a) {
        if (parent[a] == a) return a;

        return parent[a] = find(parent[a]);
    }

    public static void union(int rootA, int rootB) {
        parent[rootA] = rootB;
    }
}